README.md
# Knot Home

A small micropub and microsub client and feed reader, which is part of a larger system called [Knot System](https://github.com/knot-system). You can install it as a standalone service, or use the [Knot Installer](https://github.com/knot-system/knot-installer), which also installs other modules alongside it.

**This is an early beta version!** Some things may break, or change in the future!

## Initial Setup

Your server needs to run at least PHP 8.0 or later.

Copy all the files into a directory on your webserver, then open the url to this path in a webbrowser. The first time setup will run automatically. If you see no output, try adding `?debug` to the url, for example `https://www.example.com/knot-home/?debug`, this should show more information about what happens during setup.

The first time setup will create a `config.php` in the root folder that will hold the configuration of your system. This file is unique to your website and very important - keep a backup around, if you want to make sure to not lose anything.

The setup also creates some other files that are needed, like a (hidden) `.htaccess` file and a `cache/` folder. When you delete those item, they will be re-created as needed. They will also be automatically deleted and recreated when you make a system update.

You can now open the url in a webbrowser, and log in with a micropub and/or microsub enabled site.

## Additional Options

You may want to edit the `config.php` in the root directory a bit after the initial setup and add additional settings:

```php
<?php

return [
	'debug' => true, // should be true while in alpha
	'logging' => true, // should be true while in alpha; writes logfiles into the /log directory
	'theme' => 'default',
	'theme-color-scheme' => 'blue', // for the default theme, this can be 'blue' (default), 'green', 'red' or 'lilac'
	'microsub' => true, // set this to false if you want to disable microsub functionality completely
	'micropub' => true, // set this to false if you want to disable micropub functionality completely
	'cookie_lifetime' => 60*60*24*10, // 10 days, in seconds
	'cache_lifetime' => 60*60*24*30, // 30 days, in seconds

	// for more config options, see the file system/config.php
	
];

```

This section of the README will be expanded later, when we reach a stable state.

The loading order of the config is as follows:
1) `system/config.php`
   gets overwritten by:
2) `theme/{themename}/config.php` (if it exists)
   gets overwritten by:
3) `config.php` (in the root folder)

## Custom Theme

You can duplicate the `theme/default/` folder, rename it and update the theme name and author information in the `theme/{themename}/theme.php`.

You can define the theme your site uses in the `config.php` file like this:
```php
return [
	// site_title and other fields ...
	'theme' => '{themename}',
];
```

If the theme folder does not exist, the system will automatically use the *default* theme.

You can also create a `theme/{themename}/snippets/` folder and copy files from `system/site/snippets/` into this folder, to overwrite them on a per-theme basis. All the files in the `snippets/` folder have a version number at the start of the file, so you can see if they were updated since you last copied them. The auto-updater will also show you, which of the snippets in your custom theme are out of date and need updating.

The `theme/{themename}/functions.php` contains some functions that get called when the theme gets loaded.

The `theme/{themename}/config.php` can overwrite config options from `system/config.php` (but gets itself overwritten by the local `config.php` in the root directory), so the custom theme can for example set its own image sizes.

## Updating

**Important:** Befre updating, backup your `config.php` (and your custom theme inside the `theme/` folder, if you have any). Better be safe than sorry.

You can use [Knot Control](https://github.com/knot-system/knot-control) to update this module automatically. Or you use the following instructions:

Create a new empty file called `update` (or `update.txt`) in the root folder of your installation. Then open the website, and append `?update` to the URL to trigger the update process. **Important:** if you don't finish the update, manually delete the `update` (or `update.txt`) file (if the update process finishes, this file gets deleted automatically).

Follow the steps of the updater. It will show you all of the new release notes that happened since your currently installed version - read them carefully, especially at the moment, because some things will change and may need manual intervention. After the update is complete, and if you have a custom theme installed, it will list all the files you manually need to update in your custom theme - you should do so, or you may miss out on new functionality, or the site may even break completely.

After updating, open your website; this will trigger the setup process again, that creates some missing files. Then check if everything works as expected.

### manual update

If you want to perform a manual update, delete the `system/`, `theme/default/` and `cache/` folders, as well as the `index.php`, `.htaccess`, `README.md` and `changelog.txt` files from the root folder. Then download the latest (or an older) release from the releases page. Upload the `system/` and `theme/default/` folders and the `index.php`, `README.md` and `changelog.txt` file from the downloaded release zip-file into your web directory. Then open the url in a webbrowser; this will trigger the setup process again and create some missing files.

If you have a custom theme active, make sure all your snippets are up to date and at least the same version as the corresponding files inside `system/site/snippets/`.

### system reset

If you want to reset the whole system, delete the following files and folders and open the url in a webbrowser to re-trigger the setup process:
- `.htaccess`
- `config.php`
- the `cache/` folder
- maybe the custom theme folders in the `theme/` directory (leave the `theme/default/` directory there, though)

## Backup

You should routinely backup your content. To do so, copy these files & folders to a secure location:

- the `config.php`. This contains the theme you use and other settings
- if you have a custom theme inside the `theme/` directory, make a backup of it as well. The `theme/default/` theme comes with the system, so no need to back it up

When you want to restore a backup, delete the current folders & files from your webserver, and upload your backup. You should also delete the `cache/` folder, so everything gets re-cached and is up to date. If you also want to update or reset your system, see the *Update* section above.

## Relocation

You can move the system to a new host by copying all the files to the new location. You can omit the `cache/` folder, it gets recreated and repopulated automatically on the first access at the new location.
If the installation is in a (or gets moved to a) subfolder, and you change the name of the subfolder, you need to delete the `.htaccess` file, so that it gets regenerated with the correct new path.


changelog.txt
v.0.2.1

- moved repo to 'knot-system' organization on GitHub
- updated default theme to v.0.2.1


v.0.2.0

- renamed to 'Knot Home' (was 'Sekretär')
- updated default theme to v.0.2.0


v.0.1.6

- new config option 'link_preview_autorefresh'; set to 'true' to allow automatic refreshing of link previews instead of showing a 'refresh'-icon; this will result in layout shifts after the page finished loading
- periodically check, if the access_token is still valid, if not log out the user; this makes sure to not show the user as logged in without the ability to interact with the read or write screen
- if the access_token is no longer valid, but a refresh_token was provided, we now automatically get a new access_token via the refresh_token
- try to revoke access_token when logging out
- bugfix for remote preview images from YouTube
- use 'imagecopyresampled' instead of 'imagecopyresized' for better image quality
- handle link previews with UTF-8 correctly (this fixes YouTube Link Previews)
- added compatibility with homestead-control


v.0.1.5

- default theme: only show 3 link previews, additional link previews are hidden behind a 'more' button
- allow endpoint discovery via 'indieauth-metadata'
- allow endpoint discovery via http header
- better general compatibility with IndieAuth
- add_header() & remove_header() functions, to set http headers
- bugfixes & enhancements


v.0.1.4

- link previews
- updater now correctly deletes the session cookie and cache files
- new config option, 'show_item_content' (defaults to true) - set it to false, to only show link previews instead of item content
- get_config() helper function
- added basic pwa manifest, so you can add Sekretär to your home screen in 'standalone' mode
- added refresh button to read pagination
- bugfixes & enhancements


v.0.1.3

- you can now set the default page after login via the new config option 'homepage' (can be set to 'microsub' or 'micropub', defaults to 'dashboard')
- removed timeline caching, because postamt is now much faster
- sidebar now scrolls to active feed, if there is more than a viewport worth of feeds to display
- session handling updated; save all data in cachefile, only save session_id in session cookie
- setup: add comments to .htaccess, at beginning and end of our block
- bugfixes & enhancements


v.0.1.2

- read: timeline can be filtered by feed
- updater: fix for displaying relevant release information
- default theme: some design updates
- css & js files: start URL with '//' (instead of 'https://' or 'http://') to avoid mixed content
- better session handling
- don't delete .htaccess file on update
- bugfixes & enhancements


v.0.1.1

- updater: allow updating to the current (unstable) dev release
- compatibility with the homestead updater
- bugfixes and enhancements


v.0.1.0

**if you use a custom theme with snippets, you need to update all snippets**

this is the first beta release.

- very basic feedlist importer (for .txt, .json or .opml)
- login form: option to remember the URL in a cookie
- force logout all sessions on system update
- new config option 'allowed_urls'; this will restrict the login to the specified urls, if set
- refresh session lifetime on auto-login, so the users stayes logged in longer than 10 days
- renamed global $sekretaer to $core for better compatibility between modules (you need to update all snippets in a custom theme though)
- moved some functions and classes into a _shared folder, for simpler sharing between modules
- system config.php moved from system/site/config.php to system/config.php
- updater fix
- bugfixes & enhancements
- compatibility with the homestead installer


alpha.7

- micropub/write: don't require content to be set
- micropub/write: bugfix for image upload
- microsub/read: better experience when adding new feeds: feeds are automatically detected, when multiple feeds are detected you can select which one to follow, additional information about the feed are retrieved
- microsub/read: exported feedlist now has the channel name in the filename
- removed 'Sekretär' h1, and added Dashboard to navigation
- added additional HTML tags to allowed_html_elements config option: <u>, <del>, <pre>, <blockquote>, <code>, <audio>
- automatic link detection enhancements
- added robots.txt to disallow indexing
- default theme update to alpha.7
- bugfixes & enhancements


alpha.6

- micropub/write: design update
- micropub/write: form is now available as a snippet
- micropub/write: show image preview when selecting an image
- microsub/read: design update
- microsub/read: automatically select the first channel (that is not 'notifications'), if no channel is selected
- microsub/read: export feedlist as .txt, .json or .opml (experimental)
- microsub/read: manage channels (create, rename, delete)
- updated default theme to alpha.6
- the first-time setup now runs silently in the background, if everything succeeds
- bugfixes & enhancements


alpha.5

- show categories of posts, if available
- active channel is still clickable
- also show pagination before postlist
- show feed title, if available
- shorten long posts with a 'read more' button
- updated default theme to alpha.5


alpha.4

you need to logout and login again, because the scopes were updated

- microsub: added additional scopes for full support of all actions / methods
- microsub: bugfixes and enhancements


alpha.3

designupdate; following & unfollowing of feeds; basic feed pagination

- theme: moved menu to the left side
- theme: updated to alpha.3
- microsub: allow following of new feeds
- microsub: allow unfollowing of feeds
- microsub: very basic pagination (only 'next page' link is available for now)
- microsub: bugfix when switching channels
- micropub: cache tags for a short time


alpha.2

you need to manually update to alpha.2 because of a bug in the auto-updater of alpha.1

- bugfix for endless redirect, if autologin is active and the session cache file vanishes

system/config.php
<?php

// NOTE: you can overwrite these options:
// - in your custom theme, via /theme/{themename}/config.php
// - and/or via the config.php in the root folder

return [
	'debug' => false, // show additional information when an error occurs
	'logging' => true, // write logfiles into the /log directory
	'theme' => 'default',
	'theme-color-scheme' => 'default', // depends on the theme; the default theme supports 'blue' (default), 'green', 'red', 'lilac'
	'scope' => [ 'read', 'create', 'follow', 'channels', 'follow', 'mute', 'block' ],
	'microsub' => true, // set to false to disable the 'read' tab
	'micropub' => true, // set to false to disable the 'write' tab
	'homepage' => 'dashboard', // can be 'dashboard', 'microsub' or 'micropub'
	'cookie_lifetime' => 60*60*24*10, // 10 days, in seconds
	'cache_lifetime' => 60*60*24*30, // 30 days, in seconds
	'session_lifetime' => 60*60*24*10, // 10 days, in seconds
	'datetime_format' => 'Y-m-d H:i', // see this list for possible parameters: https://www.php.net/manual/en/datetime.format.php#refsect1-datetime.format-parameters
	'allowed_html_elements' => [ 'del', 'pre', 'blockquote', 'code', 'b', 'strong', 'u', 'i', 'em', 'ul', 'ol', 'li', 'p', 'br', 'span', 'a', 'img', 'video', 'audio' ], // allowed html elements for post content (in read mode), everything else (like <script> or <iframe> tags) gets stripped
	'allowed_urls' => [], // an array with urls of allowed users ('me' parameters)
	'user_agent' => 'knot/home/', // version will be automatically appended
	'image_target_width' => 800,
	'preview_target_width' => 800,
	'image_jpg_quality' => 70, // quality of jpg images; you neeed to empty the cache when changing this option
	'image_png_to_jpg' => true, // convert png images to jpg (faster, but looses transparency); you need to empty the cache when changing this option
	'image_background_color' => [ 255, 255, 255 ], // backgroundcolor for transparent images, when 'image_png_to_jpg' option is set to true; you need to empty the cache when changing this option
	'link_preview_max_age' => 60*60*6, // refresh link previews after x seconds
	'link_preview_nojs_refresh' => false, // refresh link previews via PHP; this makes pageloading with preview links slower, but does not rely on JavaScript to fetch link previews in the background
	'link_preview_autorefresh' => false, // set to true to automatically refresh link previews instead of showing a 'refresh' icon; this will result in layout shifts after the site is loaded
	'show_item_content' => true, // show content of feed items (set to true) or only preview-links (set to false)
];
